keps/sig-node/20191211-node-maintenance-lease.md

```diff
@@ -0,0 +1,175 @@
+---
+title: Node Maintenance Lease
+authors:
+  - "@michaelgugino"
+owning-sig: sig-cluster-lifecycle
+participating-sigs:
+  - sig-node
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-12-11
+last-updated: 2019-12-19
+status: provisional
+see-also:
+  - "https://github.com/kubernetes/enhancements/issues/1403"
+---
+
+# Node Maintenance Lease
+
+## Table of Contents
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories [optional]](#user-stories-optional)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+    - [Impact to the Kubelet](#impact-to-the-kubelet)
+    - [Impact to Scheduleability](#impact-to-scheduleability)
+    - [Client Behavior](#client-behavior)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Alternatives](#alternatives)
+  - [Annotations on Node or elsewhere](#annotations-on-node-or-elsewhere)
+  - [Field in Node Spec Or Status](#field-in-node-spec-or-status)
+  - [Use existing system namespace for lease object](#use-existing-system-namespace-for-lease-object)
+  - [Use optional non-system namespace](#use-optional-non-system-namespace)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+- [ ] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+## Summary
+
+Node Maintenance Leases are useful to coordinate disruptive actions
+controllers or administrators might wish to take against a node. This lease
+is a voluntary mechanism and allows coordination among different components
+without those components needing knowledge of each other.
+
+## Motivation
+
+Currently, there is no centralized way to inform a controller or user that
```

Comment:
Out of curiosity - do you know how folks are addressing this issue now? I.e. these days, if a cluster administrator needs to perform a disruptive operation against a node, what will they do?

Reply:
It depends on what components they have running.  Those components might use some annotation or other mechanism to pause operations, but I don't know of any specific component doing that today.

Some non-trivial conflicting components that exist today:

https://github.com/kubevirt/node-maintenance-operator

NMO is for assisting in scheduling maintenance of bare metal hardware.

https://github.com/openshift/machine-config-operator/

MCO is for upgrading node configurations.

You don't want those two components to compete.

In addition, say you're powered off a bare metal machine that was provisioned via cluster-api.  The Machine Health Checker https://github.com/kubernetes-sigs/cluster-api/pull/1684 would delete that node without some kind of intervention today.